be/src/storage/binlog_manager.cpp

```diff
@@ -43,9 +43,195 @@ BinlogManager::~BinlogManager() {
+    Status status = BinlogUtil::list_binlog_file_ids(_path, &binlog_file_ids);
+    if (!status.ok()) {
+        std::string errMsg = fmt::format("Failed to init binlog because of list error, tablet {}, status: {}",
+                                         _tablet_id, status.to_string());
+        LOG(ERROR) << errMsg;
+        _init_failure.store(true);
+        return Status::InternalError(errMsg);
+    }
+    binlog_file_ids.sort();
+
+    // 2. recover binlog from the largest version to the smallest version
+    std::vector<int64_t> useless_file_ids;
+    std::vector<BinlogFileMetaPBPtr> recovered_file_metas;
+    auto version_it = sorted_valid_versions.rbegin();
+    std::list<int64_t>::reverse_iterator file_id_it = binlog_file_ids.rbegin();
+    int64_t num_versions_recovered = 0;
+    while (version_it != sorted_valid_versions.rend()) {
+        int64_t version = *version_it;
+        BinlogLsn min_lsn = version == min_valid_lsn.version() ? min_valid_lsn : BinlogLsn(version, 0);
+        Status recover_status = _recover_version(version, min_lsn, binlog_file_ids, file_id_it, recovered_file_metas,
+                                                 &useless_file_ids);
+        if (!recover_status.ok()) {
+            break;
+        }
+        version_it++;
+        num_versions_recovered++;
+    }
+
+    if (version_it != sorted_valid_versions.rend()) {
+        std::string errMsg = fmt::format(
+                "Failed to init binlog because can't find binlog for all of versions, "
+                "tablet: {}, min_valid_lsn: {}, num expected versions: {}, num recovered version: {}, last failed "
+                "version: {}",
+                _tablet_id, min_valid_lsn.to_string(), sorted_valid_versions.size(), num_versions_recovered,
+                *version_it);
+        LOG(ERROR) << errMsg;
+        if (VLOG_IS_ON(3)) {
+            std::stringstream ss;
+            ss << "detail of valid versions for tablet: " << _tablet_id << ", [";
+            auto it = sorted_valid_versions.begin();
+            ss << *it;
+            it++;
+            while (it != sorted_valid_versions.end()) {
+                ss << ", " << *it;
+                it++;
+            }
+            ss << "]";
+            VLOG(3) << ss.str();
+        }
+        _init_failure.store(true);
+        return Status::InternalError(errMsg);
+    }
+
+    // 3. construct metas according to the recovery result
+    std::reverse(recovered_file_metas.begin(), recovered_file_metas.end());
+    BinlogBuildResult build_result;
+    build_result.next_file_id = binlog_file_ids.empty() ? BINLOG_MIN_FILE_ID : (*binlog_file_ids.rbegin() + 1);
+    build_result.metas = recovered_file_metas;
+    _apply_build_result(&build_result);
+
+    // 4. cleanup useless binlog files
+    while (file_id_it != binlog_file_ids.rend()) {
+        useless_file_ids.push_back(*file_id_it);
+        file_id_it++;
+    }
+    std::reverse(useless_file_ids.begin(), useless_file_ids.end());
```

Comment:
sorry, why so many reverse

Reply:
file ids are pushed back to `useless_file_ids` in descending order, and we want to delete them in ascending order, so traverse the vector in the reverse order